rust/dialog-storage/src/storage/backend/s3/access.rs

```diff
@@ -0,0 +1,705 @@
+/// Does NOT include bucket, endpoint, or session duration - those are
+/// specified when creating a session.
+#[derive(Debug, Clone)]
+pub enum Service {
+    /// AWS S3 or S3-compatible service (including Cloudflare R2).
+    S3 {
+        /// AWS region (e.g., "us-east-1", "auto" for R2)
+        region: String,
+    },
+}
+
+impl Service {
+    /// Create S3 service configuration.
+    pub fn s3(region: impl Into<String>) -> Self {
+        Self::S3 {
+            region: region.into(),
+        }
+    }
+
+    /// Get the region for this service.
+    fn region(&self) -> &str {
+        match self {
+            Self::S3 { region } => region,
+        }
+    }
+
+    /// Get the service name for signing.
+    fn name(&self) -> &str {
+        match self {
+            Self::S3 { .. } => "s3",
+        }
+    }
+}
+
+/// S3 Access Control List (ACL) settings.
+///
+/// These are canned ACLs supported by S3 and S3-compatible services.
+/// See: https://docs.aws.amazon.com/AmazonS3/latest/userguide/acl-overview.html#canned-acl
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Acl {
+    /// Owner gets FULL_CONTROL. No one else has access rights.
+    Private,
+    /// Owner gets FULL_CONTROL. The AllUsers group gets READ access.
+    PublicRead,
+    /// Owner gets FULL_CONTROL. The AllUsers group gets READ and WRITE access.
+    PublicReadWrite,
+    /// Owner gets FULL_CONTROL. The AuthenticatedUsers group gets READ access.
+    AuthenticatedRead,
+    /// Object owner gets FULL_CONTROL. Bucket owner gets READ access.
+    BucketOwnerRead,
+    /// Both the object owner and the bucket owner get FULL_CONTROL.
+    BucketOwnerFullControl,
+}
+
+impl Acl {
+    /// Get the S3 ACL header value.
+    pub fn as_str(&self) -> &'static str {
+        match self {
+            Self::Private => "private",
+            Self::PublicRead => "public-read",
+            Self::PublicReadWrite => "public-read-write",
+            Self::AuthenticatedRead => "authenticated-read",
+            Self::BucketOwnerRead => "bucket-owner-read",
+            Self::BucketOwnerFullControl => "bucket-owner-full-control",
+        }
+    }
+}
+
+/// Pre-derived signing key for request signing.
+#[derive(Debug, Clone)]
+struct SessionKey(Vec<u8>);
+
+impl SessionKey {
+    fn new(credentials: &Credentials, service: &Service, date: &[u8]) -> Self {
+        let secret = format!("AWS4{}", credentials.secret_access_key);
+        // Derive signing key: HMAC chain of date -> region -> service -> "aws4_request"
+        let k_date = Self::hmac(secret.as_bytes(), date);
+        let k_region = Self::hmac(&k_date, service.region().as_bytes());
+        let k_service = Self::hmac(&k_region, service.name().as_bytes());
+        Self(Self::hmac(&k_service, b"aws4_request"))
+    }
+
+    fn hmac(key: &[u8], data: &[u8]) -> Vec<u8> {
+        let mut mac = Hmac::<Sha256>::new_from_slice(key).expect("HMAC accepts any key size");
```

Comment:
In general I do not like to ship code that may panic at runtime. If it's infallible in practice, an inline comment justifying the panic-inducing call should be included.